### ISSUE TITLE
Report case sensitivity issues

### ISSUE THREAD

**Issue by WilcoFiers** [dequelabs/axe-core#2695]
Saw a question come up today about:

```html
<ul role="list">
  <li role="listItem">...</li>
</ul>
```

The problem is that capital letter "I" in "listItem". There are two possible issues there:

1. I'm not so sure this should be case sensitive, we should test our ARIA rules to figure out where case sensitivity matters and where it doesn't. If it followed the specs, it should be case insensitive in HTML and case sensitive in XHTML.

2. In such cases, it would be good for axe-core to explicitly call out that the item was failed because of case sensitivity.

3. It would be nice to report this particular case as needs review instead of a fail, because the implicit role here is the correct role. This is a false positive.

**Comment by WilcoFiers** [comment]
May be enough to typecast the result of getRole(). We'll need to test if this works reliably.

**Comment by jnurthen** [comment]
> I'm not so sure this should be case sensitive, we should test our ARIA rules to figure out where case sensitivity matters and where it doesn't. If it followed the specs, it should be case insensitive in HTML and case sensitive in XHTML.

This is defined in https://w3c.github.io/html-aria/#case-sensitivity - it MUST be lower case only

**Comment by dylanb** [comment]
@WilcoFiers close?

**Comment by WilcoFiers** [comment]
This is an ongoing discussion, see https://github.com/w3c/html-aria/issues/280, https://github.com/w3c/html-aria/pull/287 and https://github.com/w3c/aria/issues/1264.

I ran the following through a few AT:

```html
<button role="heading">lower-case</button>
<button role="HEADING">upper-case</button>
```

The latest versions of all the AT/browser combinations axe-core supports treats the role as case insensitive. That includes Safari+VO, JAWS+IE, NVDA+Firefox, NVDA+Chrome. I found it wasn't working in my older version of Firefox, but after upgrading to the latest one, there was no issue.

In conclusion, axe-core should normalise to lower case when testing anything to do with the role attribute. When this issue was opened, I still think this was an issue, but now that Firefox has changed I think we can reasonably say that using upper case in roles doesn't matter.

**Comment by jnurthen** [comment]
@dylanb looking at the extension code it should be ok 
I see the line 		`role = role ? role.toLowerCase() : "";`

**Comment by dylanb** [comment]
@wilco, if Dragon does in fact support all cases, then I think we should put this into the category of `warning`

**Comment by WilcoFiers** [comment]
We don't really have a "warning" category. If something works, I'm not sure why would need to tell someone about it. Seems like that'd only create more noise.

**Comment by jnurthen** [comment]
> Just confirmed with @mfairchild365 that dragon is case sensitive. The following does not work in Dragon
> 
> ```
> <div role="Button">Capital B</div>
> <div role="BUTTON">All caps</div>
> ```
> 
> > * Dragon does this via document.querySelectorAll (case sensitive attribute values by default) but apparently this would work if Nuance decided to finally update their code… document.querySelectorAll("[role=button i]")
> > * Dragon will discover the element if it has an onclick attribute and then parse out that it is a button no matter what case was used (this threw me through a loop at first because I tested by adding onclick to all the buttons and it somehow worked)

Looking at the code this looks correct. I was only looking at the clickable code branch not the non-clickable code branch previously.

**Comment by dan-tripp** [comment]
I've been working on this and I could use some guidance. 

One hurdle is to make each check's CSS selector case-insensitive.  I couldn't find a case-insensitive CSS selector option.  So I used parser.registerAttrEqualityMods() to make a generic regex CSS selector option, then I used the regex "i" flag to get case-insensitivity.

For example, this lets me change [page-no-duplicate-banner.json](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/keyboard/page-no-duplicate-banner.json) from this:
`"selector": "header:not([role]), [role=banner]",`
to this:
`"selector": "header:not([role]), [role/='/banner/i']",`

It works.  But is it the cleanest way?

**Comment by straker** [comment]
@dan-tripp Alright. Talked with @WilcoFiers and we decided that we just want to deal with informing users of the case sensitivity problem and deal with axe-core being case insensitive in a separate issue.

So what we'll want to do is update the [`invalidrole` check](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/aria/invalidrole-evaluate.js#L31) to call `toLowerCase` on the role before it passes it to `isValidRole` (that will stop `aria-roles` from failing on the roles). We'll then want to add a new check to the `aria-roles` rule that specifically looks for roles that are not all lower case and flag them item as Needs Review (i.e. returns `undefined`) and informs that user that not all lower case roles are not widely supported.

**Comment by dan-tripp** [comment]
Right on.  I think that I can do all this. Hard to say when, but I will try.

**Comment by straker** [comment]
Nah, we'll do it when we're ready

**Comment by dan-tripp** [comment]
Okay.

**Comment by dan-tripp** [comment]
I'd like to note something for completeness: 

Unfortunately I am very unlikely to ever finish the work that I started on this issue.

Reason being: I work at Siteimprove now.  If I contributed to axe-core while working at Siteimprove, it would look a little weird.

**Comment by lsprr** [comment]
> @dan-tripp Alright. Talked with @WilcoFiers and we decided that we just want to deal with informing users of the case sensitivity problem and deal with axe-core being case insensitive in a separate issue.
> 
> So what we'll want to do is update the [`invalidrole` check](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/aria/invalidrole-evaluate.js#L31) to call `toLowerCase` on the role before it passes it to `isValidRole` (that will stop `aria-roles` from failing on the roles). We'll then want to add a new check to the `aria-roles` rule that specifically looks for roles that are not all lower case and flag them as Needs Review (i.e. returns `undefined`) and informs the user that roles that are not all lower case are not widely supported.

Hello @straker, may I work on this issue? If not, please let me know. Thanks.

**Comment by straker** [comment]
@lsprr yep! Please let us know if you have any questions.

**Comment by padmavemulapati** [comment]
Verified with the latest dev branch code base for axe-core,
When role is upper case valid role like `<div role='COMMAND' id='aria-rolesfail1'>fail</div>` - earlier it is failing and now it is passing
Also for uppercase nonsensical role like '<div id="target" role="FOO">Contents</div>` - this is failing even now.

https://github.com/dequelabs/axe-core/assets/47409579/dd41711b-a473-48ed-9ab7-6c636e86c515

## Environment:
| **Label**                        | **Value** |
| -------------------------- | ------ |
| **Product**               | _axe-core_ |
|**Version** | _4.9.0(Dev Branch Code base_ |
| **OS-Details**     | _MAC- Intel Core i7 - 11.6.8 _ |
|**BrowserDetails**|_Chrome Version 124.0.6367.119 (Official Build) (64-bit)_|
